### PR TITLE
Issue 382

### LINKED ISSUE #382: Incorrect code generation with broadcast and merge
This results in incorrect code:

```
  int numel_x = 3;
  int numel_y = 4;
  int numel_z = 5;

  auto tv0 = makeConcreteTensor({numel_x, numel_y});
  fusion.addInput(tv0);

  auto tv1 = add(tv0, new Float(1));
  auto tv2 = broadcast(tv1, {false, false, true});
  auto tv3 = makeConcreteTensor({numel_x, numel_y, numel_z});
  fusion.addInput(tv3);
  auto tv4 = add(tv2, tv3);
  fusion.addOutput(tv4);

  // Works fine without merging but generates invalid code otherwise
  tv2->merge(1);
  tv4->merge(1);

  tv1->computeAt(tv4, 1);
```

See `testGPU_FusionBCastMerge` in https://github.com/naoyam/pytorch/blob/bug-bcast-and-merge/test/cpp/jit/test_gpu.cpp.

The generated code is:

```
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 3> T3, Tensor<float, 3> T4) {
  for(size_t i106 = 0; i106 < 3; ++i106) {
    float T2[(4 * 1)];
    float T1[4];
    for(size_t i109 = 0; i109 < 4; ++i109) {
      T1[i109]
        = T0[(i106 * T0.stride[0]) + (i109 * T0.stride[1])]
        + float(1);
    }
    for(size_t i113 = 0; i113 < (4 * 1); ++i113) {
      if (((i113 / 5) < 4)) {
        T2[(i113 / 5)]
           = T1[(i113 / 5)];
      }
    }
    for(size_t i118 = 0; i118 < (4 * 5); ++i118) {
      if ((((i118 / 5) < 4) && ((i118 % 5) < 5))) {
        T4[(i106 * T4.stride[0]) + i118]
          = T2[(i118 / 5)]
          + T3[(i106 * T3.stride[0]) + ((i118 / 5) * T3.stride[1]) + ((i118 % 5) * T3.stride[2])];
      }
    }
  }
}
```

The index of each of the two `T2` expressions should not be divided by `5`. Similarly, when `T1` is used for `T2`, its index should be just `i113` without `/ 5`. 

Merging the `T4` axes is fine. Merging `T2` broadcast dimension seems problematic. Merging broadcast dimensions may not make much sense. Is it considered an illegal operation?

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #382 

When traversing a computeAt chain, all root domains are mapped between tensor pairs. This isn't strictly correct, and is the reason of the indexing bug of #382. This PR only maps CA root domains.

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -9275,6 +9275,51 @@ TEST(NVFuserTest, Issue329_CUDA) {
...
+  fusion.addInput(tv0);
+
+  auto tv1 = add(tv0, new Float(1));
+  auto tv2 = broadcast(tv1, {false, false, true});
+  auto tv3 = makeSymbolicTensor(3);
+  fusion.addInput(tv3);
+  auto tv4 = add(tv2, tv3);
+  fusion.addOutput(tv4);
+
+  tv2->merge(1);
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:9292]: Does this statement do anything?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 2]
```diff
@@ -657,11 +657,36 @@ generateIndexAndExtentMap(
...
+      std::unordered_set<Val*> consumer_CA_root_vals = IterVisitor::getInputsTo(
+          std::vector<Val*>(consumer_CA_ids.begin(), consumer_CA_ids.end()));
+
+      std::unordered_set<IterDomain*> consumer_CA_root_ids;
+      for (auto val : consumer_CA_root_vals) {
+        if (val->getValType().value() == ValType::IterDomain) {
+          consumer_CA_root_ids.emplace(val->as<IterDomain>());
+        }
+      }
+#if 0
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Please clean this up

**Review by csarofeen — APPROVED** [review]
Looks good, just minor cleanup needed.

File: test/cpp/jit/test_gpu.cpp (hunk 1 above)
Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp:9292]: Yes. The bug of issue #382 is caused by this merge.

**Pull request merged**